<a href="https://colab.research.google.com/github/AIWalaBro/Capstone_Project_ML/blob/main/Houses_Preprocessing/Houses_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)

In [3]:
import warnings

# Suppress the SettingWithCopyWarning
warnings.filterwarnings("ignore")



In [4]:
df = pd.read_csv('/content/houses.csv')
df.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69

In [5]:
df.shape

(1044, 21)

In [6]:
# know the info
# check null
# check duplicated
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1044 entries, 0 to 1043
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    1044 non-null   object
 1   link             1044 non-null   object
 2   society          453 non-null    object
 3   price            968 non-null    object
 4   rate             1005 non-null   object
 5   area             1044 non-null   object
 6   areaWithType     987 non-null    object
 7   bedRoom          987 non-null    object
 8   bathroom         987 non-null    object
 9   balcony          987 non-null    object
 10  additionalRoom   589 non-null    object
 11  address          1031 non-null   object
 12  noOfFloor        967 non-null    object
 13  facing           674 non-null    object
 14  agePossession    987 non-null    object
 15  nearbyLocations  913 non-null    object
 16  description      1036 non-null   object
 17  furnishDetails   743 non-null    

In [7]:
df.isnull().sum()

property_name        0
link                 0
society            591
price               76
rate                39
area                 0
areaWithType        57
bedRoom             57
bathroom            57
balcony             57
additionalRoom     455
address             13
noOfFloor           77
facing             370
agePossession       57
nearbyLocations    131
description          8
furnishDetails     301
features           370
rating             137
property_id          8
dtype: int64

In [8]:
df.duplicated().sum()

13

# observation:-
- we found most of the columns are null values
- we found duplicated values
- most of the columns are in the object but you have to chnage it into float or int values

In [9]:
# lets drop the duplicates
df = df.drop_duplicates()

In [10]:
# now lets check once again shape of the data
df.shape

(1031, 21)

# Observation:-
- rows are reduced 1044 - 1031 = 13

In [11]:
df.columns

Index(['property_name', 'link', 'society', 'price', 'rate', 'area',
       'areaWithType', 'bedRoom', 'bathroom', 'balcony', 'additionalRoom',
       'address', 'noOfFloor', 'facing', 'agePossession', 'nearbyLocations',
       'description', 'furnishDetails', 'features', 'rating', 'property_id'],
      dtype='object')

In [12]:
# we have to drop unnecessary columns
df.drop(columns = ['link','property_id'], inplace = True)

In [13]:
df.shape

(1031, 19)

In [14]:
# rename columns
df.rename(columns = {'rate':'price_per_sqft'}, inplace = True)

In [15]:
df.head(2)

property_name           society  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  Bptp Visionnaire   
1  5 Bedroom House for sale in Sector 21A Faridabad               NaN   

        price    price_per_sqft                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   

                  areaWithType     bedRoom     bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)  5 Bedrooms  4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)  5 Bedrooms  4 Bathrooms   2 Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   

                                     address noOfFloor      facing  \
0  29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1     Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN

# Society

In [16]:
df['society'].value_counts()

International City by SOBHA Phase 2                  26
Emaar MGF Marbella                                   26
Vipul Tatvam Villa                                   24
International City by Sobha Phase 1                  23
Unitech Uniworld Resorts                             13
DLF City Plots                                       11
DLF City Plots Phase 2                               11
Unitech Espace                                       11
Eldeco Mansionz                                       9
Bptp Visionnaire                                      8
Not Applicable                                        7
Uppal Southend4.0 ★                                   7
Sushant Lok 1 Builder Floors                          6
Ansals Palam Vihar                                    6
DLF City Phase 1                                      5
Experion Windchants3.8 ★                              5
DLF Alameda                                           4
Unitech Aspen Greens                            

In [17]:
df['society'].value_counts().shape

(185,)

In [18]:
import re
df['society'] = df['society'].apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip()).str.lower()

In [19]:
df['society'].value_counts()

nan                                                  586
emaar mgf marbella                                    26
international city by sobha phase 2                   26
vipul tatvam villa                                    24
international city by sobha phase 1                   23
unitech uniworld resorts                              13
dlf city plots                                        11
dlf city plots phase 2                                11
unitech espace                                        11
eldeco mansionz                                        9
bptp visionnaire                                       8
not applicable                                         7
uppal southend                                         7
sushant lok 1 builder floors                           6
ansals palam vihar                                     6
dlf city phase 1                                       5
experion windchants                                    5
dlf alameda                    

In [20]:
df['society'].isnull().sum()

0

In [21]:
df.isnull().sum()

property_name        0
society              0
price               76
price_per_sqft      39
area                 0
areaWithType        57
bedRoom             57
bathroom            57
balcony             57
additionalRoom     453
address             13
noOfFloor           76
facing             369
agePossession       57
nearbyLocations    131
description          8
furnishDetails     299
features           369
rating             137
dtype: int64

# Observation:-
- nan values are present in the society columns but no ableto see df
- we can do one thing nan values can be replace as independent

In [22]:
df['society'] = df['society'].str.replace('nan','independent')

In [23]:
df['society'].value_counts()

independent                                          586
emaar mgf marbella                                    26
international city by sobha phase 2                   26
vipul tatvam villa                                    24
international city by sobha phase 1                   23
unitech uniworld resorts                              13
dlf city plots                                        11
dlf city plots phase 2                                11
unitech espace                                        11
eldeco mansionz                                        9
bptp visionnaire                                       8
not applicable                                         7
uppal southend                                         7
sushant lok 1 builder floors                           6
ansals palam vihar                                     6
dlf city phase 1                                       5
experion windchants                                    5
dlf alameda                    

In [24]:
df.head(3)

property_name           society  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   
1  5 Bedroom House for sale in Sector 21A Faridabad       independent   
2  10 Bedroom House for sale in Sushant Lok Phase 1       independent   

        price    price_per_sqft                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   

                                     address noOfFloor      facing  \
0  29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1     Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2      Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...

#Price

In [25]:
df['price'].value_counts()

3.5 Crore           19
5.5 Crore           17
6.5 Crore           16
6 Crore             14
5 Crore             13
4.5 Crore           13
70 Lac              13
8 Crore             13
50 Lac              13
75 Lac              13
8.5 Crore           12
7 Crore             12
85 Lac              12
1.5 Crore           11
4 Crore             11
10 Crore            10
Price on Request    10
45 Lac              10
2.25 Crore          10
2.5 Crore           10
60 Lac              10
90 Lac              10
9 Crore             10
9.5 Crore            9
12 Crore             9
80 Lac               9
1.7 Crore            9
2 Crore              9
1.1 Crore            9
1 Crore              8
6.75 Crore           8
11 Crore             8
1.6 Crore            8
4.6 Crore            8
1.3 Crore            7
95 Lac               7
12.5 Crore           7
3.75 Crore           7
1.2 Crore            7
55 Lac               7
2.1 Crore            7
65 Lac               7
40 Lac               7
1.35 Crore 

In [26]:
# we noticed that Price on Request are in the value_counts
df = df[df['price'] != 'Price on Request']

In [27]:
df.shape

(1021, 19)

#### Observation:- we noticed we left 1021 rows

In [28]:
def treat_price(x):
    if type(x) == float:
        return x
    else:
        if x[1] == 'Lac':
            return round(float(x[0])/100,2)
        else:
            return round(float(x[0]),2)

In [29]:
df['price'] = df['price'].str.split(' ').apply(treat_price)

In [30]:
df['price']

0        5.25
1        5.70
2        2.10
3        5.00
4        3.00
5        4.50
6       12.00
7       20.00
8       10.85
9        1.95
10       4.45
11       9.00
12       5.00
13      15.00
14       8.25
15       2.00
16      10.75
17       2.10
18       7.35
19       4.45
20       3.40
21      12.00
22       8.40
23       8.50
24       7.60
25       7.00
26       4.75
27       3.50
28       5.00
29      10.56
30       5.46
31      10.00
32       9.00
33      10.00
34       1.95
35       3.30
36      13.00
37       7.20
38       6.00
39       9.00
40       5.80
41       0.55
42       7.49
43      13.50
44      11.95
45      12.39
46       6.25
47        NaN
48       5.40
49       5.75
50       2.55
51       5.60
52      17.00
53       7.99
54       3.75
55       7.25
56       8.40
57       3.55
58       2.50
59       2.50
60       9.00
61       2.10
62        NaN
63        NaN
64       4.50
65       6.50
66        NaN
67        NaN
68        NaN
69        NaN
70        NaN
71    

# price_per_sqft


In [41]:
df['price_per_sqft'].value_counts()

₹ 22,222/sq.ft.     14
₹ 11,111/sq.ft.     11
₹ 33,333/sq.ft.     11
₹ 10,000/sq.ft.     10
₹ 26,667/sq.ft.      6
₹ 13,333/sq.ft.      6
₹ 37,037/sq.ft.      6
₹ 26,235/sq.ft.      6
₹ 30,556/sq.ft.      6
₹ 31,481/sq.ft.      6
₹ 9,444/sq.ft.       5
₹ 23,333/sq.ft.      5
₹ 23,148/sq.ft.      5
₹ 11,100/sq.ft.      5
₹ 8,000/sq.ft.       5
₹ 8,889/sq.ft.       5
₹ 5,000/sq.ft.       5
₹ 35,000/sq.ft.      4
₹ 32,407/sq.ft.      4
₹ 9,722/sq.ft.       4
₹ 18,889/sq.ft.      4
₹ 24,691/sq.ft.      4
₹ 25,000/sq.ft.      4
₹ 27,778/sq.ft.      4
₹ 21,605/sq.ft.      4
₹ 34,722/sq.ft.      4
₹ 6,944/sq.ft.       4
₹ 36,111/sq.ft.      3
₹ 13,889/sq.ft.      3
₹ 20,000/sq.ft.      3
₹ 16,058/sq.ft.      3
₹ 8,235/sq.ft.       3
₹ 60,000/sq.ft.      3
₹ 12,222/sq.ft.      3
₹ 8,333/sq.ft.       3
₹ 14,815/sq.ft.      3
₹ 12,000/sq.ft.      3
₹ 6,666/sq.ft.       3
₹ 15,555/sq.ft.      3
₹ 10,556/sq.ft.      3
₹ 7,500/sq.ft.       3
₹ 11,250/sq.ft.      3
₹ 15,625/sq.ft.      3
₹ 25,926/sq

In [42]:
df['price_per_sqft'] = df['price_per_sqft'].str.split('/').str.get(0).str.replace('₹'," ").str.replace(',','').str.strip().astype(float)

In [43]:
df['price_per_sqft'].dtype

dtype('float64')

In [44]:
df['price_per_sqft'].isnull().sum()
# nan values are there

29

# observation:- 29 null values are there


# Bedrooms

In [45]:
df['bedRoom'].value_counts()

4 Bedrooms     232
5 Bedrooms     194
3 Bedrooms     146
2 Bedrooms      88
6 Bedrooms      81
9 Bedrooms      45
1 Bedroom       38
8 Bedrooms      33
7 Bedrooms      31
12 Bedrooms     28
10 Bedrooms     21
16 Bedrooms     12
13 Bedrooms      4
18 Bedrooms      2
19 Bedrooms      2
11 Bedrooms      2
20 Bedrooms      1
21 Bedrooms      1
36 Bedrooms      1
34 Bedrooms      1
14 Bedrooms      1
Name: bedRoom, dtype: int64

In [35]:
# df['bedRoom'].str.get(0)

# Nan values are there

In [47]:
df[df['bedRoom'].isnull()].shape

(57, 19)

In [48]:
df = df[~df['bedRoom'].isnull()]

In [49]:
df.shape

(964, 19)

# observation:-
- we remove null values and 964 rows

In [52]:
df['bedRoom'] = df['bedRoom'].str.get(0).astype('int')

In [54]:
df['bedRoom'].shape

(964,)

In [55]:
df.shape

(964, 19)

#Bathroom

In [56]:
df['bathroom'].value_counts()

4 Bathrooms     228
3 Bathrooms     165
5 Bathrooms     137
2 Bathrooms     114
6 Bathrooms      82
1 Bathroom       65
9 Bathrooms      42
7 Bathrooms      39
8 Bathrooms      28
12 Bathrooms     22
10 Bathrooms     11
16 Bathrooms      8
18 Bathrooms      4
13 Bathrooms      4
11 Bathrooms      4
20 Bathrooms      3
17 Bathrooms      3
14 Bathrooms      2
21 Bathrooms      1
36 Bathrooms      1
34 Bathrooms      1
Name: bathroom, dtype: int64

In [57]:
df['bathroom'].isnull().sum()

0

In [61]:
df['bathroom'] = df['bathroom'].str.split(' ').str.get(0).astype('int')

In [62]:
df['bathroom']

0        4
1        4
2       10
3       21
4       12
5       20
6        5
7       16
8        6
9        4
10       3
11       7
12      13
13       6
14       6
15       4
16       9
17       7
18       4
19       6
20       9
21       6
22       4
23       4
24       9
25       5
26       5
27       9
28       9
29       5
30       5
31       6
32      10
33       4
34       2
35       4
36      16
37       3
38       4
39       5
40       5
41       1
42      12
43      16
44      12
45      18
46       5
47       5
48       5
49       4
50       3
51       6
52       8
53       6
54       6
55       4
56       5
57       4
58       4
59       4
60       5
61       3
64       5
65       4
79      12
80       4
90       4
91       4
101      4
106      3
107      3
108      5
109      6
112      5
113      2
114      4
116      5
117      3
118     13
119      5
120      4
121      2
122      4
123      2
124      3
125      3
126      4
127      5
128      5
129      6
130      5

In [63]:
df.head()

property_name           society  price  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   5.25   
1  5 Bedroom House for sale in Sector 21A Faridabad       independent   5.70   
2  10 Bedroom House for sale in Sushant Lok Phase 1       independent   2.10   
3    21 Bedroom House for sale in Sector 54 Gurgaon       independent   5.00   
4  12 Bedroom House for sale in Sushant Lok Phase 1       independent   3.00   

   price_per_sqft                   area                 areaWithType  \
0         20115.0  (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   
1        105751.0   (50 sq.m.) Plot Area   Plot area 539(50.07 sq.m.)   
2         38251.0   (51 sq.m.) Plot Area       Plot area 61(51 sq.m.)   
3         43066.0  (108 sq.m.) Plot Area  Plot area 129(107.86 sq.m.)   
4         53763.0   (52 sq.m.) Plot Area    Plot area 62(51.84 sq.m.)   

   bedRoom  bathroom       balcony                      additionalRoom  \
0        5         4  3+ Balconies                        Servant Room   
1        5         4   2 Balconies  Store Room,Pooja Room,Servant Room   
2        1        10  3+ Balconies                        Servant Room   
3        2        21  3+ Balconies                        Servant Room   
4        1        12  3+ Balconies                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...

In [64]:
df.shape

(964, 19)

#Balcony

In [65]:
df['balcony'].value_counts()

3+ Balconies    365
2 Balconies     220
3 Balconies     159
No Balcony      115
1 Balcony       105
Name: balcony, dtype: int64

In [69]:
df['balcony'] = df['balcony'].str.split(' ').str.get(0).str.replace('No', '0')

In [70]:
df['balcony']

0       3+
1        2
2       3+
3       3+
4       3+
5       3+
6       3+
7       3+
8        1
9       3+
10      3+
11      3+
12      3+
13      3+
14      3+
15       3
16      3+
17      3+
18      3+
19       3
20       3
21      3+
22       3
23       2
24       3
25      3+
26       1
27      3+
28      3+
29       3
30      3+
31       3
32       3
33      3+
34       2
35       2
36      3+
37       2
38      3+
39       3
40      3+
41       1
42      3+
43      3+
44      3+
45      3+
46       3
47      3+
48      3+
49       2
50       2
51      3+
52      3+
53       3
54       3
55       2
56      3+
57      3+
58      3+
59      3+
60      3+
61       2
64       1
65      3+
79      3+
80       2
90      3+
91      3+
101     3+
106      2
107      2
108      2
109      3
112     3+
113      1
114      2
116     3+
117      2
118     3+
119      3
120     3+
121      2
122      3
123      2
124      2
125      2
126      2
127     3+
128      2
129      3
130      3

In [71]:
df.head(2)

property_name           society  price  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   5.25   
1  5 Bedroom House for sale in Sector 21A Faridabad       independent   5.70   

   price_per_sqft                   area                 areaWithType  \
0         20115.0  (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   
1        105751.0   (50 sq.m.) Plot Area   Plot area 539(50.07 sq.m.)   

   bedRoom  bathroom balcony                      additionalRoom  \
0        5         4      3+                        Servant Room   
1        5         4       2  Store Room,Pooja Room,Servant Room   

                                     address noOfFloor      facing  \
0  29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1     Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN

#Additional room

In [73]:
df['additionalRoom'].value_counts()

Servant Room                                     81
Pooja Room,Study Room,Servant Room,Store Room    60
Others                                           58
Pooja Room                                       38
Pooja Room,Study Room,Servant Room               33
Pooja Room,Study Room,Servant Room,Others        32
Store Room                                       31
Pooja Room,Servant Room                          24
Study Room                                       19
Study Room,Servant Room                          19
Servant Room,Store Room                          15
Pooja Room,Study Room                            11
Pooja Room,Store Room                            11
Pooja Room,Others                                11
Servant Room,Pooja Room                           8
Servant Room,Others                               8
Pooja Room,Store Room,Study Room,Servant Room     8
Pooja Room,Servant Room,Store Room                7
Store Room,Servant Room                           6
Study Room,S

In [74]:
df['additionalRoom'].isnull().sum()

386

# Observation:-
- 386 null values present are there
- so what we can do replace null values with not available

In [75]:
df['additionalRoom'].fillna('not available', inplace =True)

In [76]:
df['additionalRoom'] = df['additionalRoom'].str.lower()

In [78]:
df['additionalRoom'].isnull().sum()

0

#Floors

In [81]:
df['noOfFloor'].value_counts()

3 Floors     304
2 Floors     288
1 Floors     181
4 Floors     136
5 Floors       9
0 Floors       6
14 Floors      3
6 Floors       3
11 Floors      2
12 Floors      2
10 Floors      2
16 Floors      1
51 Floors      1
41 Floors      1
27 Floors      1
20 Floors      1
13 Floors      1
21 Floors      1
02 Floors      1
33 Floors      1
Name: noOfFloor, dtype: int64

In [82]:
df['noOfFloor'].isnull().sum()

19

#Observation:-
- 19 null values present over there

In [86]:
df['noOfFloor'] = df['noOfFloor'].str.split(' ').str.get(0)

In [87]:
df['noOfFloor'].shape

(964,)

#### we will handle 19 null values later on

In [88]:
df.rename(columns={'noOfFloor':'floorNum'},inplace=True)

In [90]:
df.head(1)

property_name           society  price  \
0  5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   5.25   

   price_per_sqft                   area                 areaWithType  \
0         20115.0  (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   

   bedRoom  bathroom balcony additionalRoom  \
0        5         4      3+   servant room   

                                     address floorNum      facing  \
0  29b, Sector 70A Gurgaon, Gurgaon, Haryana        3  North-East   

     agePossession                                    nearbyLocations  \
0  0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   

                                         description  \
0  Do you wish to buy an independent house in sec...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...

#Facing

In [91]:
df['facing'].isnull().sum()

303

In [92]:
df['facing'].value_counts()

East          178
North-East    158
North         109
West           75
North-West     44
South-East     39
South          34
South-West     24
Name: facing, dtype: int64

In [93]:
df['facing'].fillna('NA',inplace=True)

In [94]:
df['facing'].isnull().sum()

0

- null values are filled

# Area

In [96]:
df['area'].value_counts()

(301 sq.m.) Plot Area        43
(84 sq.m.) Plot Area         37
(251 sq.m.) Plot Area        31
(167 sq.m.) Plot Area        25
(420 sq.m.) Plot Area        24
(226 sq.m.) Plot Area        21
(418 sq.m.) Plot Area        20
(125 sq.m.) Plot Area        19
(201 sq.m.) Plot Area        18
(50 sq.m.) Plot Area         16
(180 sq.m.) Plot Area        16
(135 sq.m.) Plot Area        15
(209 sq.m.) Plot Area        14
(171 sq.m.) Plot Area        12
(293 sq.m.) Plot Area        11
(334 sq.m.) Plot Area        11
(33 sq.m.) Plot Area         11
(42 sq.m.) Plot Area         11
(264 sq.m.) Plot Area        10
(220 sq.m.) Plot Area        10
(51 sq.m.) Plot Area          9
(145 sq.m.) Plot Area         8
(502 sq.m.) Plot Area         8
(151 sq.m.) Plot Area         8
(136 sq.m.) Plot Area         7
(134 sq.m.) Plot Area         7
(161 sq.m.) Plot Area         7
(100 sq.m.) Plot Area         6
(242 sq.m.) Plot Area         6
(5 sq.m.) Plot Area           6
(836 sq.m.) Plot Area         6
(52 sq.m

In [99]:
# df['price']*10000000/
df['price_per_sqft']

0        20115.0
1       105751.0
2        38251.0
3        43066.0
4        53763.0
5        39062.0
6        37037.0
7        48889.0
8        38150.0
9        10077.0
10       32963.0
11       27778.0
12       34083.0
13       33200.0
14       30556.0
15       14914.0
16       29861.0
17       23333.0
18       41762.0
19       15451.0
20       12938.0
21       38095.0
22       25926.0
23       26235.0
24       38760.0
25       25926.0
26        9374.0
27       13410.0
28       21124.0
29       32593.0
30       24170.0
31       37037.0
32       28490.0
33      122222.0
34       13542.0
35       19097.0
36       48148.0
37       22222.0
38       22222.0
39       33333.0
40       32222.0
41        7290.0
42       27741.0
43       50000.0
44       43534.0
45       45889.0
46       34722.0
47           NaN
48       20690.0
49       29716.0
50       16378.0
51       17284.0
52       37777.0
53       12519.0
54       25562.0
55       27971.0
56       21000.0
57       24500.0
58        7058

In [105]:
df['area'] = round((df['price']*10000000)/df['price_per_sqft'])

#Observation:-
- NaN values are present

In [97]:
df['area'].isnull().sum()

0

In [100]:
df.insert(loc=1,column='property_type',value='house')

In [101]:
df.head()

property_name property_type  \
0    5 Bedroom House for sale in Sector 70A Gurgaon         house   
1  5 Bedroom House for sale in Sector 21A Faridabad         house   
2  10 Bedroom House for sale in Sushant Lok Phase 1         house   
3    21 Bedroom House for sale in Sector 54 Gurgaon         house   
4  12 Bedroom House for sale in Sushant Lok Phase 1         house   

            society  price  price_per_sqft                   area  \
0  bptp visionnaire   5.25         20115.0  (242 sq.m.) Plot Area   
1       independent   5.70        105751.0   (50 sq.m.) Plot Area   
2       independent   2.10         38251.0   (51 sq.m.) Plot Area   
3       independent   5.00         43066.0  (108 sq.m.) Plot Area   
4       independent   3.00         53763.0   (52 sq.m.) Plot Area   

                  areaWithType  bedRoom  bathroom balcony  \
0  Plot area 290(242.48 sq.m.)        5         4      3+   
1   Plot area 539(50.07 sq.m.)        5         4       2   
2       Plot area 61(51 sq.m.)        1        10      3+   
3  Plot area 129(107.86 sq.m.)        2        21      3+   
4    Plot area 62(51.84 sq.m.)        1        12      3+   

                       additionalRoom  \
0                        servant room   
1  store room,pooja room,servant room   
2                        servant room   
3                        servant room   
4                              others   

                                       address floorNum      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana        3  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana        2          NA   
2        Sushant Lok Phase 1, Gurgaon, Haryana        5        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana        5       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana        5        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...

In [102]:
df.shape

(964, 20)

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 964 entries, 0 to 1043
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    964 non-null    object 
 1   property_type    964 non-null    object 
 2   society          964 non-null    object 
 3   price            945 non-null    float64
 4   price_per_sqft   945 non-null    float64
 5   area             964 non-null    object 
 6   areaWithType     964 non-null    object 
 7   bedRoom          964 non-null    int64  
 8   bathroom         964 non-null    int64  
 9   balcony          964 non-null    object 
 10  additionalRoom   964 non-null    object 
 11  address          959 non-null    object 
 12  floorNum         945 non-null    object 
 13  facing           964 non-null    object 
 14  agePossession    964 non-null    object 
 15  nearbyLocations  848 non-null    object 
 16  description      964 non-null    object 
 17  furnishDetails 

In [108]:
df.to_csv('house_cleaned.csv',index=False)

In [106]:
df['area']

0          2610.0
1           539.0
2           549.0
3          1161.0
4           558.0
5          1152.0
6          3240.0
7          4091.0
8          2844.0
9          1935.0
10         1350.0
11         3240.0
12         1467.0
13         4518.0
14         2700.0
15         1341.0
16         3600.0
17          900.0
18         1760.0
19         2880.0
20         2628.0
21         3150.0
22         3240.0
23         3240.0
24         1961.0
25         2700.0
26         5067.0
27         2610.0
28         2367.0
29         3240.0
30         2259.0
31         2700.0
32         3159.0
33          818.0
34         1440.0
35         1728.0
36         2700.0
37         3240.0
38         2700.0
39         2700.0
40         1800.0
41          754.0
42         2700.0
43         2700.0
44         2745.0
45         2700.0
46         1800.0
47            NaN
48         2610.0
49         1935.0
50         1557.0
51         3240.0
52         4500.0
53         6382.0
54         1467.0
55        

In [107]:
df['area'].isnull().sum()

19